In [25]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:

pd.options.display.max_columns = None

# DATA SOURCE


DrivenData. (2015). Pump it Up: Data Mining the Water Table. Retrieved [December 3 2024] from https://www.drivendata.org/competitions/7/pump-it-up-data-mining-the-water-table.

# Data Underdstanding

In [3]:
#Test Data
test_set_values_df = pd.read_csv("data/Test_set_values.csv")
test_set_values_df

id  amount_tsh date_recorded                  funder  gps_height  \
0      50785         0.0    2013-02-04                    Dmdd        1996   
1      51630         0.0    2013-02-04  Government Of Tanzania        1569   
2      17168         0.0    2013-02-01                     NaN        1567   
3      45559         0.0    2013-01-22              Finn Water         267   
4      49871       500.0    2013-03-27                  Bruder        1260   
...      ...         ...           ...                     ...         ...   
14845  39307         0.0    2011-02-24                  Danida          34   
14846  18990      1000.0    2011-03-21                    Hiap           0   
14847  28749         0.0    2013-03-04                     NaN        1476   
14848  33492         0.0    2013-02-18                 Germany         998   
14849  68707         0.0    2013-02-13  Government Of Tanzania         481   

        installer  longitude   latitude                 wpt_name  num_private  \
0            DMDD  35.290799  -4.059696  Dinamu Secondary School            0   
1             DWE  36.656709  -3.309214                  Kimnyak            0   
2             NaN  34.767863  -5.004344           Puma Secondary            0   
3      FINN WATER  38.058046  -9.418672           Kwa Mzee Pange            0   
4          BRUDER  35.006123 -10.950412          Kwa Mzee Turuka            0   
...           ...        ...        ...                      ...          ...   
14845          Da  38.852669  -6.582841                Kwambwezi            0   
14846        HIAP  37.451633  -5.350428         Bonde La Mkondoa            0   
14847         NaN  34.739804  -4.585587                  Bwawani            0   
14848         DWE  35.432732 -10.584159                 Kwa John            0   
14849  Government  34.765054 -11.226012         Kwa Mzee Chagala            0   

                         basin   subvillage   region  region_code  \
0                     Internal       Magoma  Manyara           21   
1                      Pangani      Kimnyak   Arusha            2   
2                     Internal        Msatu  Singida           13   
3      Ruvuma / Southern Coast   Kipindimbi    Lindi           80   
4      Ruvuma / Southern Coast      Losonga   Ruvuma           10   
...                        ...          ...      ...          ...   
14845              Wami / Ruvu        Yombo    Pwani            6   
14846                  Pangani      Mkondoa    Tanga            4   
14847                 Internal       Juhudi  Singida           13   
14848               Lake Nyasa  Namakinga B   Ruvuma           10   
14849               Lake Nyasa        Kamba   Ruvuma           10   

       district_code            lga          ward  population public_meeting  \
0                  3          Mbulu        Bashay         321           True   
1                  2   Arusha Rural      Kimnyaki         300           True   
2                  2  Singida Rural          Puma         500           True   
3                 43         Liwale       Mkutano         250            NaN   
4                  3         Mbinga  Mbinga Urban          60            NaN   
...              ...            ...           ...         ...            ...   
14845              1       Bagamoyo         Yombo          20           True   
14846              7        Kilindi       Mvungwe        2960           True   
14847              2  Singida Rural       Ughandi         200           True   
14848              2   Songea Rural      Maposeni         150           True   
14849              3         Mbinga    Mbamba bay          40           True   

                   recorded_by scheme_management                scheme_name  \
0      GeoData Consultants Ltd        Parastatal                        NaN   
1      GeoData Consultants Ltd               VWC             TPRI pipe line   
2      GeoData Consultants Ltd               VWC                         

In [4]:
# Data to be used
Training_set_values_df = pd.read_csv("data/Training_set_values.csv")
Training_set_values_df

id  amount_tsh date_recorded           funder  gps_height  \
0      69572      6000.0    2011-03-14            Roman        1390   
1       8776         0.0    2013-03-06          Grumeti        1399   
2      34310        25.0    2013-02-25     Lottery Club         686   
3      67743         0.0    2013-01-28           Unicef         263   
4      19728         0.0    2011-07-13      Action In A           0   
...      ...         ...           ...              ...         ...   
59395  60739        10.0    2013-05-03  Germany Republi        1210   
59396  27263      4700.0    2011-05-07      Cefa-njombe        1212   
59397  37057         0.0    2011-04-11              NaN           0   
59398  31282         0.0    2011-03-08            Malec           0   
59399  26348         0.0    2011-03-23       World Bank         191   

          installer  longitude   latitude              wpt_name  num_private  \
0             Roman  34.938093  -9.856322                  none            0   
1           GRUMETI  34.698766  -2.147466              Zahanati            0   
2      World vision  37.460664  -3.821329           Kwa Mahundi            0   
3            UNICEF  38.486161 -11.155298  Zahanati Ya Nanyumbu            0   
4           Artisan  31.130847  -1.825359               Shuleni            0   
...             ...        ...        ...                   ...          ...   
59395           CES  37.169807  -3.253847   Area Three Namba 27            0   
59396          Cefa  35.249991  -9.070629     Kwa Yahona Kuvala            0   
59397           NaN  34.017087  -8.750434               Mashine            0   
59398          Musa  35.861315  -6.378573                Mshoro            0   
59399         World  38.104048  -6.747464       Kwa Mzee Lugawa            0   

                         basin    subvillage       region  region_code  \
0                   Lake Nyasa      Mnyusi B       Iringa           11   
1                Lake Victoria       Nyamara         Mara           20   
2                      Pangani       Majengo      Manyara           21   
3      Ruvuma / Southern Coast    Mahakamani       Mtwara           90   
4                Lake Victoria    Kyanyamisa       Kagera           18   
...                        ...           ...          ...          ...   
59395                  Pangani      Kiduruni  Kilimanjaro            3   
59396                   Rufiji      Igumbilo       Iringa           11   
59397                   Rufiji     Madungulu        Mbeya           12   
59398                   Rufiji        Mwinyi       Dodoma            1   
59399              Wami / Ruvu  Kikatanyemba     Morogoro            5   

       district_code             lga              ward  population  \
0                  5          Ludewa          Mundindi         109   
1                  2       Serengeti             Natta         280   
2                  4       Simanjiro           Ngorika         250   
3                 63        Nanyumbu          Nanyumbu          58   
4                  1         Karagwe        Nyakasimbi           0   
...              ...             ...               ...         ...   
59395              5             Hai  Masama Magharibi         125   
59396              4          Njombe            Ikondo          56   
59397              7         Mbarali           Chimala           0   
59398              4        Chamwino      Mvumi Makulu           0   
59399              2  Morogoro Rural        Ngerengere         150   

      public_meeting              recorded_by scheme_management  \
0               True  GeoData Consultants Ltd               VWC   
1                NaN  GeoData Consultants Ltd             Other   
2               True  GeoData Consultants Ltd               VWC   
3               True  GeoData Consultants Ltd               VWC   
4               True  GeoData Consultants Ltd               NaN   
...              ...                      ...               ... 

In [5]:
# Predictor y label
Training_set_labels_df = pd.read_csv("data/Training_set_labels.csv")
Training_set_labels_df

id    status_group
0      69572      functional
1       8776      functional
2      34310      functional
3      67743  non functional
4      19728      functional
...      ...             ...
59395  60739      functional
59396  27263      functional
59397  37057      functional
59398  31282      functional
59399  26348      functional

[59400 rows x 2 columns]

In [6]:
# Merge Dataset
# Merging the train x data to its y predictor
df_merge = pd.merge(Training_set_labels_df,Training_set_values_df,how= "inner", on = "id")

In [7]:
df_merge.sample(20)

id    status_group  amount_tsh date_recorded  \
2632   16091      functional         0.0    2013-03-14   
44165  33647  non functional         0.0    2011-08-22   
11772  44940      functional      6000.0    2013-11-03   
3275    9421      functional         0.0    2013-02-18   
16811  56444      functional       200.0    2013-02-11   
22967  70585      functional         0.0    2011-04-03   
19861  56635      functional         0.0    2011-03-07   
8131   19550      functional         0.0    2011-03-14   
27028  26861      functional         0.0    2013-03-21   
46461  27515  non functional         0.0    2012-10-15   
41713   3094      functional         0.0    2013-02-26   
8798   48680      functional      1000.0    2011-03-15   
9813   11992      functional         0.0    2013-02-15   
33664  25233      functional         0.0    2011-08-07   
7249   43444      functional         0.0    2011-08-01   
35737  41584  non functional         0.0    2011-08-03   
17556  73731      functional         0.0    2011-07-18   
36836  23419      functional         0.0    2011-07-28   
25024  71465      functional         0.0    2013-02-26   
35003  44670      functional       500.0    2013-05-04   

                       funder  gps_height  installer  longitude  latitude  \
2632                Community        1428  Community  36.752677 -3.364673   
44165                     Ilo        1748        DWE  31.764551 -8.439754   
11772          Oikos E.Afrika        1459  Community  36.840786 -3.158458   
3275                 Kalitasi        1235   Kalitasi  38.007295 -4.261005   
16811                   Norad         984        DWE  29.652329 -4.755085   
22967                     NaN           0        NaN  33.879027 -9.286732   
19861                   Commu           0      TANAS  35.858553 -6.340715   
8131                     Wsdp         647        DWE  38.462292 -5.061802   
27028             Peter Tesha        1508        DWE  37.386954 -3.257300   
46461                   Dwssp           0        DWE  33.589309 -3.709155   
41713  Government Of Tanzania        1603        DWE  37.434150 -3.288185   
8798                Rc Church        2148  RC CHURCH  34.538074 -9.320533   
9813       Private Individual        1264  Juma Maro  34.350940 -1.638997   
33664                   Hifab           0     Hesawa  33.453952 -2.862311   
7249   Government Of Tanzania           0        DWE  30.672980 -1.249584   
35737  Government Of Tanzania           4        DWE  39.054001 -5.051323   
17556                       P           0          P  33.012711 -2.396463   
36836                  Hesawa           0        DWE  30.781913 -1.409435   
25024             Netherlands           0        DWE  33.557076 -2.872734   
35003             Aco/germany        1300        DWE  37.313780 -3.242803   

                      wpt_name  num_private            basin     subvillage  \
2632   Ambureni Primary School            0          Pangani       Ambureni   
44165       Kwa Julius Nazalio            0       Lake Rukwa        Maswepa   
11772        Pentecoste Nkuuny            0         Internal         Nkuuny   
3275                   Msikiti            0          Pangani           Vuje   
16811        Kwa Yusuf Fadhili            0  Lake Tanganyika          Mlati   
22967                  Lupando            0       Lake Nyasa        Lupando   
19861   Mvumi Makulu Secondary            0           Rufiji       Miembeni   
8131                  Kwaponga            0          Pangani       Misufini   
27028           Kwa Mzee Joram            0          Pangani        Ngaruma   
46461                Kwa Timbe            0         Internal        Italaga   
41713                 Kwa Maro            0          Pangani       Mrahembe   
8798               Kwa Lusanyo            0       Lake Nyasa        Kibaoni   
9813                 Juma Maro            0    Lake Victoria          Senta   
33664                     none            0    Lake Victoria      N

In [8]:
# Getting basic information on the merged dataset
df_merge.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 59400 entries, 0 to 59399
Data columns (total 41 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   id                     59400 non-null  int64  
 1   status_group           59400 non-null  object 
 2   amount_tsh             59400 non-null  float64
 3   date_recorded          59400 non-null  object 
 4   funder                 55765 non-null  object 
 5   gps_height             59400 non-null  int64  
 6   installer              55745 non-null  object 
 7   longitude              59400 non-null  float64
 8   latitude               59400 non-null  float64
 9   wpt_name               59400 non-null  object 
 10  num_private            59400 non-null  int64  
 11  basin                  59400 non-null  object 
 12  subvillage             59029 non-null  object 
 13  region                 59400 non-null  object 
 14  region_code            59400 non-null  int64  
 15  di

In [9]:
# Checking the columns available in our dataset
df_merge.columns

Index(['id', 'status_group', 'amount_tsh', 'date_recorded', 'funder',
       'gps_height', 'installer', 'longitude', 'latitude', 'wpt_name',
       'num_private', 'basin', 'subvillage', 'region', 'region_code',
       'district_code', 'lga', 'ward', 'population', 'public_meeting',
       'recorded_by', 'scheme_management', 'scheme_name', 'permit',
       'construction_year', 'extraction_type', 'extraction_type_group',
       'extraction_type_class', 'management', 'management_group', 'payment',
       'payment_type', 'water_quality', 'quality_group', 'quantity',
       'quantity_group', 'source', 'source_type', 'source_class',
       'waterpoint_type', 'waterpoint_type_group'],
      dtype='object')

In [10]:
# finding the shape of our dataset
df_merge.shape

(59400, 41)

In [11]:
# finding unique values in each column

object_columns = df_merge.select_dtypes(include=['object','int64','float64']).columns

object_column_unique = {col : df_merge[col].nunique() for col in object_columns}

columns_unique_df =pd.DataFrame(list(object_column_unique.items()), columns= ['column_name', 'number_of_unique_values'])
columns_unique_df


column_name  number_of_unique_values
0                      id                    59400
1            status_group                        3
2              amount_tsh                       98
3           date_recorded                      356
4                  funder                     1897
5              gps_height                     2428
6               installer                     2145
7               longitude                    57516
8                latitude                    57517
9                wpt_name                    37400
10            num_private                       65
11                  basin                        9
12             subvillage                    19287
13                 region                       21
14            region_code                       27
15          district_code                       20
16                    lga                      125
17                   ward                     2092
18             population                     1049
19         public_meeting                        2
20            recorded_by                        1
21      scheme_management                       12
22            scheme_name                     2696
23                 permit                        2
24      construction_year                       55
25        extraction_type                       18
26  extraction_type_group                       13
27  extraction_type_class                        7
28             management                       12
29       management_group                        5
30                payment                        7
31           payment_type                        7
32          water_quality                        8
33          quality_group                        6
34               quantity                        5
35         quantity_group                        5
36                 source                       10
37            source_type                        7
38           source_class                        3
39        waterpoint_type                        7
40  waterpoint_type_group                        6

## Basic understanding of each of the columns based on their cardinality

**High Cardinality columns**
- id (59400) = Unique values are equal to the total number of rows of dataframe indicating its the unique identifier of the dataset
- longitude and latitude = indicates that the wells are spread on a huge geographical region
- wpt_name = indicates a large variety on water point names
- sub_villages = indicates wells are located on a high number of unique sub-villages

**Medium Cardinality columns**

- funder = indicates a huge number of funders i.e people who funded the project
- gps_height = indicates a large number of wells were built along multiple altitude heights
- ward -Indicates wells are found in multiple different wards
- installer - indicates that there are multiple organizations that have constructed wells in Tanzania
- scheme_name - Indicates there are a lot of different water schemes in the area
- amount_tsh (98)-  this represents the total static head this means the elevation between the free level of water till the discharge point of the pump

**Low Cardinality Columns**
- status group = indicates the three types of status a well might be i.e functional, non-functional, functional but in need of repair
- public meeting (2) = Indicates a boolean value where its either a yes or no
- management_group (5) = Indicates there are 5 unique ways a well may be managed
- permit (2) = shows a well can be built having a permit or no permit
- source_class = shows the classification of the source of the water



## numerical columns that are distinct
- id
- longitude and latitude
- region code 
- district code
- construction year


In [12]:
df_merge.describe()

id     amount_tsh    gps_height     longitude      latitude  \
count  59400.000000   59400.000000  59400.000000  59400.000000  5.940000e+04   
mean   37115.131768     317.650385    668.297239     34.077427 -5.706033e+00   
std    21453.128371    2997.574558    693.116350      6.567432  2.946019e+00   
min        0.000000       0.000000    -90.000000      0.000000 -1.164944e+01   
25%    18519.750000       0.000000      0.000000     33.090347 -8.540621e+00   
50%    37061.500000       0.000000    369.000000     34.908743 -5.021597e+00   
75%    55656.500000      20.000000   1319.250000     37.178387 -3.326156e+00   
max    74247.000000  350000.000000   2770.000000     40.345193 -2.000000e-08   

        num_private   region_code  district_code    population  \
count  59400.000000  59400.000000   59400.000000  59400.000000   
mean       0.474141     15.297003       5.629747    179.909983   
std       12.236230     17.587406       9.633649    471.482176   
min        0.000000      1.000000       0.000000      0.000000   
25%        0.000000      5.000000       2.000000      0.000000   
50%        0.000000     12.000000       3.000000     25.000000   
75%        0.000000     17.000000       5.000000    215.000000   
max     1776.000000     99.000000      80.000000  30500.000000   

       construction_year  
count       59400.000000  
mean         1300.652475  
std           951.620547  
min             0.000000  
25%             0.000000  
50%          1986.000000  
75%          2004.000000  
max          2013.000000

- amount_total_static_head = the min amount is 0, the mean value is 317.65. There is also an indication that majority of the values are 0 because the 1st quartile and the median being 0
- gps_height = min value is -90 might indicate the wells are located below sea level
- population = min amount is 0. This might be an error or indicates that the area where the wells are have no poulation while the maximum is 30500

# Data Cleaning


In [13]:
# checking the number of null values in my dataset

df_merge.isna().sum()

id                           0
status_group                 0
amount_tsh                   0
date_recorded                0
funder                    3635
gps_height                   0
installer                 3655
longitude                    0
latitude                     0
wpt_name                     0
num_private                  0
basin                        0
subvillage                 371
region                       0
region_code                  0
district_code                0
lga                          0
ward                         0
population                   0
public_meeting            3334
recorded_by                  0
scheme_management         3877
scheme_name              28166
permit                    3056
construction_year            0
extraction_type              0
extraction_type_group        0
extraction_type_class        0
management                   0
management_group             0
payment                      0
payment_type                 0
water_qu

There are missing value in the following columns
- funder 
- installer 
- sub_village
- public_meeting
- scheme_management
- scheme_name
- permit

All this are categorical columns and missing values might be caused by missed information in the data collection process


In [14]:
# checking the information contained on the columns with the missing values

missing_cols = df_merge.columns[df_merge.isnull().any()]

df_merge.loc[: ,missing_cols].sample(20)

funder           installer  \
38614             Kilindi District Co                 DWE   
52606                 Concern/governm  Concern/Government   
25597                            Wsdp                 DWE   
25270                          Hesawa              Hesawa   
29469                        Nethalan                 RWE   
8420                              Isf                 ISF   
56854          Government Of Tanzania                 DWE   
33180                            Dwsp                 DWE   
12252                              Is                  Is   
43928                               0                   0   
1538                            Rwssp              GEOTAN   
12                       World Vision        World vision   
54028                       Kiliwater           Kiliwater   
11951                     Netherlands                 DWE   
10625                             Bgm                 BGM   
50060  Rural Water Supply And Sanitat                 DWE   
26571                    World Vision              VTECOS   
17773          Government Of Tanzania          Government   
31416                          Danida              DANIDA   
40745                           Oxfam                 DWE   

                subvillage public_meeting scheme_management  \
38614           Kidereko A           True               VWC   
52606               Itambo           True               VWC   
25597         Nyantugutu A           True               WUG   
25270             Kichongo           True               WUG   
29469                Ndumi           True               VWC   
8420                Mbueni           True   Water authority   
56854             Olubomba           True               VWC   
33180              Mwabuku            NaN               WUG   
12252             Ugindoni           True               VWC   
43928  Mtaa Wa Kipunguni B            NaN  Private operator   
1538               Magungu           True               WUG   
12                 Nkilifa           True               WUG   
54028               Wama B           True           Company   
11951         Mwang'Honoli           True               WUG   
10625                Iboja           True               NaN   
50060              Misheni            NaN        Parastatal   
26571               Ikunya           True               VWC   
17773              Geterer           True               VWC   
31416              Godawni           True               VWC   
40745           Ieletulele           True               VWC   

                             scheme_name permit  
38614                                NaN  False  
52606                           Ijindile   True  
25597                                NaN  False  
25270                                NaN    NaN  
29469                           Mvaji Ri   True  
8420                              Saseni   True  
56854                 Olkokola pipe line   True  
33180                                NaN  False  
12252                                NaN   True  
43928                                NaN  False  
1538                                None   True  
12                                   NaN  False  
54028                EKTM 3 water supply   True  
11951                                NaN  False  
10625                                NaN   True  
50060                                NaN   True  
26571  NCHULOWAIBALE WATER SUPPLY SCHEME   True  
17773                                NaN   True  
31416     Libango water use group scheme   True  
40745                  Nabaiye pipe line   True

In [15]:
# for categorical columns that explain extra information on the well we can fill null values by the word unknown i.e funder, installer , subvillage, scheme_management, scheme_name

df_merge[['funder', 'installer', 'subvillage','scheme_management','scheme_name']] = df_merge[['funder', 'installer', 'subvillage','scheme_management','scheme_name']].fillna("unknown")


In [ ]:
# for the remaining lets check the value counts as they are boolean values i.e between two choices

object_columns = df_merge.columns[df_merge.isna().any()]

plt.subplot(2,1, figsize=(10,8))

for 



{'public_meeting': True     51011
 False     5055
 Name: public_meeting, dtype: int64,
 'permit': True     38852
 False    17492
 Name: permit, dtype: int64}